ai_edge_torch/generative/layers/experimental/attention.py

```diff
@@ -174,10 +172,9 @@ def forward(
     k = torch.cat((k_roped, k[..., n_elem:]), dim=-1)
 
     if kv_cache is not None:
-      k_cache, v_cache = kv_cache
-      k = k_cache.index_copy_(1, input_pos, k)
-      v = v_cache.index_copy_(1, input_pos, v)
-      kv_cache = k, v
+      kv_cache = kv_utils.update(kv_cache, input_pos, k, v, use_hlfb=self.enable_hlfb)
```

Comment:
I was surprised by this, in that the logic is to decide on a per-inference basis which function is called (the HLFB version or not). I would have thought that the function to call would be set as an attribute on the module at `__init__` time, so that it would just look like:

```
kv_cache = self.cache_update(kv_cache, input_pos, k, v)
```

where `self.cache_update` is set via something like:

```
self.cach_update = kv_utils._update_kv_hlfb_impl if enable_hlfb else kv_utils._update_kv_base_impl
```

If it traces and exports the same either way then ultimately it's not a big deal, just seems a bit odd to essentially do a dynamic function dispatch at forward time. You can resolve if you have considered this and discarded the idea, but curious to know what you think.

Reply:
I did think about it a bit when implementing this. Originally, I had `update` as a part of `EKVCache` object but that did not make sense since we are doing an out of place update. Hence `update` became a stand alone function within `ekv_cache` module.

The option were then to either define two variant of `update` within that modules and make one of them the class kv update function (similar to how sdpa is implemented right now) or have a single variant that hides the details. The first approach would require having an if-else in all the model classes that use kv_update, i.e., 
```
    self.sdpa_func = (
        scaled_dot_product_attention_with_hlfb
        if enable_hlfb
        else scaled_dot_product_attention
    )
```
That's why I went with a single `update` function that resolves to the correct one based on the input arguments. At the end of the day, the exported program looks the same and it's more of a matter of how clean the authoring story is.

Reply:
Happy to resolve for now if you thought it through and came to a difference conclusion. Can revisit later if needed.